src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,187 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
-                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength - newChange.Span.Length;
-                            oldIndex++;
-                            newIndex++;
-                            goto nextNewChange;
-                        }
-                        else
-                        {
-                            // delete as much from old change as new change can
-                            // a new change deletion is a reduction in the old change insertion
-                            var oldChangeReduction = Math.Min(oldChange.NewLength, newChange.Span.Length);
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength - oldChangeReduction));
-                            oldDelta = oldDelta - oldChange.Span.Length + (oldChange.NewLength - oldChangeReduction);
-                            oldIndex++;
-
-                            // deduct the amount removed from oldChange from newChange's deletion span (since its already been applied)
-                            newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newChange.Span.Length - oldChangeReduction), newChange.NewLength);
-                            goto nextOldChange;
-                        }
+
+                        // adjust the new change as follows:
+                        // 1. add the old change's deletion to its deletion
+                        // 2. subtract the old change's insertion from its deletion
+                        // 3. add oldDelta to its start position so that its adjusted start position is the same as the oldChange that we removed
+                        //    - note that this can result in the newChange having a negative Start, but this gets adjusted back into a non-negative Start in the next iteration.
+                        var newDeletion = newChange.Span.Length + oldChange.Span.Length - oldChange.NewLength;
+                        newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newDeletion, validateStart: false), newChange.NewLength);
+                        if (tryGetNext(oldChanges, ref oldIndex, out oldChange)) continue;
+                        else break;
                     }
                 }
-                else
-                {
-                    // no more old changes, just add adjusted new change
-                    var adjustedNewChange = new TextChangeRange(new TextSpan(newChange.Span.Start - oldDelta, newChange.Span.Length), newChange.NewLength);
-                    AddRange(list, adjustedNewChange);
-                    newIndex++;
-                    goto nextNewChange;
-                }
             }
-            else
+
+            // there may be remaining old changes or remaining new changes (not both)
+            Debug.Assert(oldIndex == oldChanges.Length || newIndex == newChanges.Length);
```

Comment:
skipped the rest until here.

consider making this a runtime assert, not a debug one.  this should be a cheap check and we want it to fire if we got anything wrong.

Reply:
because both lists were non-empty (you check that at the start), i think you shoudl have an additional assert here of:

Debug.Assert(oldIndex < oldChanges.Length || newIndex < newChanges.Length);

i.e. at least one of these should not be at teh end.

Reply:
Have made a change to this effect (throwing in retail builds, and verifying the additional case)